egs/sad_rats/s5/local/get_speech_targets.py

```diff
@@ -0,0 +1,153 @@
+# Copyright 2020  John Morgan (ARL)
+# Apache 2.0
+
+# This script prepares targets for whole recordings for training a Speech
+# Activity detector system.
+# It takes as input a data dir where it is assumed that MFCC features have
+# already been extracted.
+# It also takes an RTTM file  with speech segments.
+# It uses these segments to obtain per-frame targets for the recordings in the format:
+# [ silence speech ]
+
+from __future__ import division
+
+import argparse
+import logging
+import numpy as np
+import subprocess
+import sys
+import itertools
+from collections import defaultdict
+
+sys.path.insert(0, 'steps')
+import libs.common as common_lib
+
+
+def get_args():
+    parser = argparse.ArgumentParser(
+        description="""This script prepares targets for whole recordings for training
+            an speech activity detector system. It just takes as input a data dir
+            where it is assumed that MFCC features have already been
+            extracted. It also takes an RTTM file containing
+            speech segments. It uses these segments to 
+            obtain per-frame targets for the recordings in the format:
+            [ silence speech ]
+        """)
+
+    parser.add_argument("--frame-shift", type=float, default=0.01,
+                        help="Frame shift value in seconds")
+    parser.add_argument("--label-smoothing", type=float, default=0.0,
+                        help="Value between 0 and 1. Amount of label smoothing to apply"
+                        "to get soft labels instead of one-hot labels")
+    parser.add_argument("reco2num_frames", type=str,
+                        help="""The number of frames per reco
+                        is used to determine the num-rows of the output matrix
+                        """)
+    parser.add_argument("rttm", type=str,
+                        help="Input RTTM file containing speech segments")
+    parser.add_argument("out_targets_ark", type=str,
+                        help="""Output archive to which the
+                        recording-level matrix will be written in text
+                        format""")
+
+    args = parser.parse_args()
+
+    if args.frame_shift < 0.0001 or args.frame_shift > 1:
+        raise ValueError("--frame-shift should be in [0.0001, 1]; got {0}"
+                         "".format(args.frame_shift))
+    return args
+
+class Segment:
+    """Stores all information about a segment"""
+    reco_id = ''
+    spk_id = ''
+    start_time = 0
+    dur = 0
+    end_time = 0
+
+    def __init__(self, reco_id, start_time, dur = None, end_time = None, label = None):
+        self.reco_id = reco_id
+        self.start_time = start_time
+        if (dur is None):
+            self.end_time = end_time
+            self.dur = end_time - start_time
+        else:
+            self.dur = dur
+            self.end_time = start_time + dur
+        self.label = label
+
+def groupby(iterable, keyfunc):
```

Comment:
Again, this is something I have realized recently. Since we are only using this function at a single place, we don't really need it. Instead, you can remove this and use `itertools.groupby()` in line 108. Note that `itertools.groupby()` needs a sorted iterable (which is why this function was created in the first place, so instead of simply passing `segments`, you would need to do `sorted(segments, key=lambda x: x.recoid)`.

Reply:
Done